pkg/server/container_execsync.go

```diff
@@ -131,50 +140,40 @@ func (c *criContainerdService) execInContainer(ctx context.Context, id string, o
 		}
 	})
 
-	// Get containerd event client first, so that we won't miss any events.
-	// TODO(random-liu): Add filter to only subscribe events of the exec process.
-	// TODO(random-liu): Use `Wait` after is fixed. (containerd#1279, containerd#1287)
-	cancellable, cancel := context.WithCancel(ctx)
-	eventstream, err := c.eventService.Subscribe(cancellable, &events.SubscribeRequest{})
-	if err != nil {
-		return nil, fmt.Errorf("failed to subscribe event stream: %v", err)
-	}
-	defer cancel()
-
+	resCh := make(chan execResult, 1)
+	go func() {
+		// Wait will return if context is cancelled.
+		exitCode, err := process.Wait(ctx)
+		resCh <- execResult{
+			exitCode: exitCode,
+			err:      err,
+		}
+		glog.V(2).Infof("Exec process %q exits with exit code %d and error %v", execID, exitCode, err)
+	}()
 	if err := process.Start(ctx); err != nil {
 		return nil, fmt.Errorf("failed to start exec %q: %v", execID, err)
 	}
 
-	exitCode, err := c.waitContainerExec(eventstream, id, execID)
-	if err != nil {
-		return nil, fmt.Errorf("failed to wait for exec in container %q to finish: %v", id, err)
+	var timeoutCh <-chan time.Time
```

Comment:
I've tried that. The problem is that once context is cancelled, we could not use it to call `Delete` or `Kill`. Of course, we could use `context.Background()` for the those operations, but I'm not sure what the code style should be.

That is also why I filed https://github.com/kubernetes-incubator/cri-containerd/issues/142.